eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxPublisher.scala

```diff
@@ -73,485 +75,270 @@ object ReplaceableTxPublisher {
+        // We wait for our parent to stop us: when that happens we will unlock utxos.
+        Behaviors.same
+      case WrappedCurrentBlockCount(currentBlockCount) =>
+        // We avoid a herd effect whenever a new block is found.
+        timers.startSingleTimer(CheckFeeKey, CheckFee(currentBlockCount), (1 + Random.nextLong(nodeParams.maxTxPublishRetryDelay.toMillis)).millis)
+        Behaviors.same
+      case CheckFee(currentBlockCount) =>
+        val currentFeerate = getFeerate(nodeParams.onChainFeeConf.feeEstimator, cmd.txInfo.confirmBefore, BlockHeight(currentBlockCount))
+        val targetFeerate_opt = if (cmd.txInfo.confirmBefore.toLong <= currentBlockCount + 6) {
+          log.debug("{} confirmation target is close (in {} blocks): bumping fees", cmd.desc, cmd.txInfo.confirmBefore.toLong - currentBlockCount)
+          // We make sure we increase the fees by at least 20% as we get close to the confirmation target.
+          Some(currentFeerate.max(tx.feerate * 1.2))
+        } else if (tx.feerate * 1.2 <= currentFeerate) {
+          log.debug("{} confirmation target is in {} blocks: bumping fees", cmd.desc, cmd.txInfo.confirmBefore.toLong - currentBlockCount)
+          Some(currentFeerate)
+        } else {
+          log.debug("{} confirmation target is in {} blocks: no need to bump fees", cmd.desc, cmd.txInfo.confirmBefore.toLong - currentBlockCount)
+          None
+        }
+        targetFeerate_opt.foreach(targetFeerate => {
+          // We check whether our currently published transaction is confirmed: if it is, it doesn't make sense to bump
+          // the fee, we're just waiting for enough confirmations to report the result back.
+          context.pipeToSelf(bitcoinClient.getTxConfirmations(tx.signedTx.txid)) {
+            case Success(Some(confirmations)) if confirmations > 0 => Stay
+            case _ => BumpFee(targetFeerate)
           }
-      }
+        })
+        Behaviors.same
+      case BumpFee(targetFeerate) => fundReplacement(replyTo, cmd, targetFeerate, txMonitor, tx)
+      case Stay => Behaviors.same
       case Stop =>
-        timeLocksChecker ! TxTimeLocksMonitor.Stop
-        Behaviors.stopped
+        txMonitor ! MempoolTxMonitor.Stop
+        unlockAndStop(cmd.input, Seq(tx.signedTx))
     }
   }
 
-  def fund(replyTo: ActorRef[TxPublisher.PublishTxResult], cmd: TxPublisher.PublishReplaceableTx, targetFeerate: FeeratePerKw): Behavior[Command] = {
-    context.pipeToSelf(addInputs(cmd.txInfo, targetFeerate, cmd.commitments)) {
-      case Success((fundedTx, fee)) => SignFundedTx(fundedTx, fee)
-      case Failure(reason) => FundingFailed(reason)
-    }
+  // Fund a replacement transaction because our previous attempt seems to be stuck in the mempool.
+  def fundReplacement(replyTo: ActorRef[TxPublisher.PublishTxResult], cmd: TxPublisher.PublishReplaceableTx, targetFeerate: FeeratePerKw, previousTxMonitor: ActorRef[MempoolTxMonitor.Command], previousTx: FundedTx): Behavior[Command] = {
```

Comment:
My first intuition would have been to handle by default the case where we have multiple versions of the tx at the same time, instead of serializing all the steps. Would that make any sense?

Reply:
I started with something like that, but it was much harder to reason about, I couldn't find a simple enough design...

Reply:
I gave it a try on branch [tx-publisher-deadline-refacto](https://github.com/ACINQ/eclair/commits/tx-publisher-deadline-refacto), WDYT? Amazingly, the tests still pass. I was able to drop the reference to `txMonitor`, not sure why you kept it around?

Reply:
> I was able to drop the reference to txMonitor, not sure why you kept it around?

We do need to stop the previous `txMonitor` when we receive a `Stop` message, right?
But I may have been too strict/overzealous here, once we stop ourselves this will automatically stop our child, so maybe this mechanism is overkill?

Reply:
> I gave it a try on branch tx-publisher-deadline-refacto, WDYT?

I think it's more complex, because you can end up with more than two transactions pending at the same time, so it becomes harder to analyze...I like restricting this as much as possible, it's already a complex mechanism, making all the boundaries explicit make it easier in my opinion to check that the behavior is correct.

Reply:
I re-visited that after https://github.com/ACINQ/eclair/pull/2113/commits/25fa5e60396c16df77f2a0fedb431d283c4a54ca now that we don't need to keep the references to the mempool tx monitors, but I still think my version is clearer.

I like that the `wait` state is the "nominal" state, where we have a transaction in the mempool and only do two things:

- wait for its status to change (confirmed or evicted)
- wait for new blocks to decide whether we should initiate RBF

Then every stage of the RBF process is a short-lived state, that should take us back to the `wait` state quickly.
Handling multiple txs in the `wait` state makes it harder to reason about the possible edge cases where we would have more than 2 txs in the list (which in practice should never happen but wouldn't be enforced at the type level).

Reply:
> I think it's more complex, because you can end up with more than two transactions pending at the same time

My gut feeling is that this could happen due to race conditions or other unforeseen circumstances and it's safer to handle it natively, but I could be wrong. And it probably wouldn't be that risky; at worse we could forget to unlock utxos I suppose. Let's leave it like that 👍